File: Adapter.py
"""
Adapter
Transform interfaces so that otherwise incompatible interfaces work together.
"""

from abc import ABC, abstractmethod


class Adaptee:
    def original_method(self):
        return "old"

class Target(ABC):
    @abstractmethod
    def adapted_method(self) -> str:
        pass

class Adapter(Target):
    def __init__(self, adaptee):
        self.adaptee:Adaptee = adaptee

    def adapted_method(self):
        old_value = self.adaptee.original_method()
        return f"{old_value}=>new"


if __name__ == "__main__":
    adaptee = Adaptee()
    target = Adapter(adaptee)

    value = target.adapted_method()
    print(value)

File: Adapter.java
/*****
 * Adapter
 * Transform interfaces so that otherwise incompatible interfaces work together.
*****/


class Adaptee { public String originalMethod() { return "old"; } }

interface Target { String adaptedMethod(); }

class Adapter_ implements Target {
    private final Adaptee adaptee;

    public Adapter_(Adaptee adaptee) { this.adaptee = adaptee; }

    @Override
    public String adaptedMethod() {
        String oldValue = adaptee.originalMethod();
        return oldValue + "=>" + "new";
    }
}


public class Adapter {
    public static void main(String[] args) {
        Adaptee adaptee = new Adaptee();
        Target target = new Adapter_(adaptee);

        String value = target.adaptedMethod();
        System.out.println(value);
    }
}
